### ISSUE TITLE
BUG: Input array broadcast error when th_badframes = 0 

### ISSUE THREAD

**Issue by evanpc** [MouseLand/suite2p#930]
### Describe the issue:

Hi, I’m sorry if I’m missing something simple, but it seems anytime I have my ops th_badframes = 0, I get an error in ROI detection. 

I’ve tried many different ops settings and many different datasets (including ones that worked well previously).

The error is always the same and persists even when all of the ops are default except tau = 0.7, fs = 30, and th_badframes = 0

### Reproduce the code example:

```python
** Found 3 tifs - converting to binary **
time 1.25 sec. Wrote 1200 frames per binary for 1 planes
>>>>>>>>>>>>>>>>>>>>> PLANE 0 <<<<<<<<<<<<<<<<<<<<<<
NOTE: not registered / registration forced with ops['do_registration']>1
      (no previous offsets to delete)
NOTE: applying default /Users/evancampbell/.suite2p/classifiers/classifier_user.npy
----------- REGISTRATION
Reference frame, 15.78 sec.
Registered 400/1200 in 7.46s
Registered 800/1200 in 14.98s
Registered 1200/1200 in 22.37s
----------- Total 45.80 sec
----------- ROI DETECTION
Binning movie in chunks of length 21
>>>ERROR<<<
Traceback (most recent call last):
  File "/Users/evancampbell/miniconda3/envs/suite2p/lib/python3.8/runpy.py", line 194, in _run_module_as_main
>>>ERROR<<<
    return _run_code(code, main_globals, None,
  File "/Users/evancampbell/miniconda3/envs/suite2p/lib/python3.8/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "/Users/evancampbell/miniconda3/envs/suite2p/lib/python3.8/site-packages/suite2p/__main__.py", line 78, in <module>
    main()
  File "/Users/evancampbell/miniconda3/envs/suite2p/lib/python3.8/site-packages/suite2p/__main__.py", line 71, in main
    run_s2p(ops, db)
  File "/Users/evancampbell/miniconda3/envs/suite2p/lib/python3.8/site-packages/suite2p/run_s2p.py", line 443, in run_s2p
    op = run_plane(op, ops_path=ops_path)
  File "/Users/evancampbell/miniconda3/envs/suite2p/lib/python3.8/site-packages/suite2p/run_s2p.py", line 296, in run_plane
    ops = pipeline(f_reg, f_raw, f_reg_chan2, f_raw_chan2, run_registration, ops, stat=stat)
  File "/Users/evancampbell/miniconda3/envs/suite2p/lib/python3.8/site-packages/suite2p/run_s2p.py", line 129, in pipeline
    ops, stat = detection.detection_wrapper(f_reg,
  File "/Users/evancampbell/miniconda3/envs/suite2p/lib/python3.8/site-packages/suite2p/detection/detect.py", line 112, in detection_wrapper
    mov = bin_movie(f_reg, bin_size, yrange=yrange, 
  File "/Users/evancampbell/miniconda3/envs/suite2p/lib/python3.8/site-packages/suite2p/detection/detect.py", line 61, in bin_movie
    mov[ik : ik + n_bins] = data
ValueError: could not broadcast input array from shape (11,492,508) into shape (0,492,508)
Interrupted by error (not finished)
```

### Version information:

suite2p v0.12.0

### Context for the issue:

I use an M1 macbook, and had previously installed and used suit2p for ~ 5-6 months (though I am still a total novice).  

I recently needed the display replaced and the machine came back with a factory reset and the newest Ventura OS installed. After re-install of iterm2, python3.8, miniconda (python 3.8), and suite2p v0.12.0, I have run into this error with every dataset I’ve tried.  

I am curious if my problem may be a software incompatibility?  

Although, I have also encountered the same error using suite2p v0.11.1 (the version I ran before my macbook was reset).

**Comment by lizard1111** [comment]
I'm getting the same input array error on about half of the datasets I've tried so far, 
They're all single-channel tiffs, exported from the same program, so I wonder why some work and others don't. 
I'm using s2p v012.0 on Ubuntu 20.04.

**Comment by jmdelahanty** [comment]
I just ran into this error as well running v0.12.0 on Debian 11. Going to try re-install/clear my conda environment in case something strange has happened after I removed and reinstalled to update.

**Comment by juhuntenburg** [comment]
Hi, 

I also get this error (the original one)
Ubuntu 20.04 , suite2p 0.12.0

We have analyzed the same dataset using the GUI but running it through the python function gives this error. 
Any help appreciated

**Comment by chriski777** [comment]
Thanks for letting us know about this. Can someone provide a dropbox link to a dataset such that this can be reproduced? We are having trouble replicating this on our end.

**Comment by jmdelahanty** [comment]
I can provide a dataset this evening, but need to pare it down to something a lot smaller. It's currently about 60k individual tiffs acquired from a Bruker 2p scope. If it still throws this error with a smaller number of tiffs I'll update this comment with the link for you. I'll also include my conda environment and the info about the machine I've run it on most recently (it's happened across several different machines).

I can also put it into H5 or whatever else you'd like if that's easier for you.

**Comment by chriski777** [comment]
Yep, that'd be great! It'd be best if it's as small as possible and can reproduce this error. Any format is fine. Thanks a bunch!

**Comment by lizard1111** [comment]
I can also upload a dataset if it might be helpful. Also, I don't know if this is a clue but sometimes it seems like registration runs fine on the initial analysis, but if I try to open the session up later on and/or tweak parameters, things "break" and I wind up with OP's error message. 

I'm a bit confused as to whether I should be saving the ops.npy from the initial run-through and reloading those ops when I reopen the session to avoid any conflicts with the default values? 

I am using the GUI to change ops values, btw.

**Comment by chriski777** [comment]
Yes, @lizard1111, if you could that would be greatly appreciated.

**Comment by jmdelahanty** [comment]
Somehow since Monday evening I have been unable to reproduce this with any of my troublesome datasets of any size... Really unsure what I did... Trying to break it again though!

**Comment by lizard1111** [comment]
That's funny, @jmdelahanty - same deal here. I've been loading in much smaller datasets than usual though, so I will see what happens with one of my regular-sized ones.

**Comment by RandallJEllis** [comment]
I am also having this issue, but like @lizard1111 it's only on some datasets. Something I noticed in the traceback is that the dimensions are somehow being changed. For example, I have a bigTIFF video, the frames of which are all 1536x2048, but my traceback shows dimensions of 1524,2038. Is this happening for anyone else? @evanpc?

```
Traceback (most recent call last):
  File "/cluster/tufts/levinlab/rellis01/calcium/suite2p_script.py", line 137, in <module>
    run_suite2p(ops, input_path, tif_files)
  File "/cluster/tufts/levinlab/rellis01/calcium/suite2p_script.py", line 110, in run_suite2p
    opsEnd=run_s2p(ops=ops,db=db)
  File "/cluster/tufts/levinlab/rellis01/condaenv/suite2p_env/lib/python3.10/site-packages/suite2p/run_s2p.py", line 446, in run_s2p
    op = run_plane(op, ops_path=ops_path)
  File "/cluster/tufts/levinlab/rellis01/condaenv/suite2p_env/lib/python3.10/site-packages/suite2p/run_s2p.py", line 296, in run_plane
    ops = pipeline(f_reg, f_raw, f_reg_chan2, f_raw_chan2, run_registration, ops, stat=stat)
  File "/cluster/tufts/levinlab/rellis01/condaenv/suite2p_env/lib/python3.10/site-packages/suite2p/run_s2p.py", line 129, in pipeline
    ops, stat = detection.detection_wrapper(f_reg,
  File "/cluster/tufts/levinlab/rellis01/condaenv/suite2p_env/lib/python3.10/site-packages/suite2p/detection/detect.py", line 112, in detection_wrapper
    mov = bin_movie(f_reg, bin_size, yrange=yrange, 
  File "/cluster/tufts/levinlab/rellis01/condaenv/suite2p_env/lib/python3.10/site-packages/suite2p/detection/detect.py", line 61, in bin_movie
    mov[ik : ik + n_bins] = data
ValueError: could not broadcast input array from shape (20,1524,2038) into shape (1,1524,2038)
```

**Comment by RandallJEllis** [comment]
I am having this issue as well, and from looking at the `bin_movie` function from [detect.py](https://github.com/MouseLand/suite2p/blob/main/suite2p/detection/detect.py) where the error is coming from, it appears to be an incompatibility between the number of frames in `f_reg`, `bin_size`, and the dimensionality of `mov`. The code below replicates the error. The error does not occur if `bin_size=10` instead of 20 as in this example, or if `f_reg` has 6040 frames instead of 6020.

```
import numpy as np
import time


f_reg = np.random.rand(6020,10,10)
badframes=None
n_frames = f_reg.shape[0]
good_frames = ~badframes if badframes is not None else np.ones(n_frames, dtype=bool)
batch_size = min(good_frames.sum(), 500)
Lyc=9
Lxc=9
xrange = (1,10)
yrange = (1,10)
num_good_frames = good_frames.sum()
bin_size=20
mov = np.zeros((num_good_frames//bin_size, Lyc, Lxc), np.float32)
ik = 0
t0 = time.time()


for k in np.arange(0, n_frames, batch_size):
	data = f_reg[k : min(k + batch_size, n_frames)]

	# exclude badframes
	good_indices = good_frames[k : min(k + batch_size, n_frames)]
	if good_indices.mean() > 0.5:
		data = data[good_indices]

	# crop to valid region
	if yrange is not None and xrange is not None:
		data = data[:, slice(*yrange), slice(*xrange)]

	# bin in time
	if data.shape[0] > bin_size:
		n_d = data.shape[0]
		data = data[:(n_d // bin_size) * bin_size]
		data = data.reshape(-1, bin_size, Lyc, Lxc).astype(np.float32).mean(axis=1)
	n_bins = data.shape[0]
	mov[ik : ik + n_bins] = data
	ik += n_bins

```

Output:
```
Traceback (most recent call last):
  File "<stdin>", line 16, in <module>
ValueError: could not broadcast input array from shape (20,9,9) into shape (1,9,9)
```

**Comment by chriski777** [comment]
Hi @RandallJEllis, thanks for providing this code! Seems this error was caused by a particular edge case in which the last batch (frames 6000-6020) happened to have the same number of frames as the `bin_size`. This error seems to have also occurred when the number of bad_frames in a given batch made it such that the filtered batch had fewer frames than `bin_size`. 

We've updated the code to cover these edge cases but without the data that caused the errors in the first place, we can't say for sure if this was the problem @lizard1111 and @jmdelahanty faced above. I'll close this for now since the commit above fixes the bug @RandallJEllis mentioned but you all can feel free to create another issue if it still persists after pulling the commit [83d4ebb](https://github.com/MouseLand/suite2p/commit/83d4ebb773778fa1655560b6c351a6999551eec9).
